apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,726 @@
-    },
-    'report-view': {
-      behaviour: 'ReportView',
-      wrapperPosition: 'left',
-      container: 'information-report'
-    }
-  },
+/**
+ * Attach a handler to run the `afterEnter` step after the page is loaded.
+ * It uses either the custom event `navigation-transition-end` or the usual
+ * `load` event.
+ * @returns {Promise} Resolved once afterEnter is run.
+ */
+function attachAfterEnterHandler() {
+  if (document.readyState === 'complete') {
+    return Promise.resolve().then(
+      () => executeNavigationStep('afterEnter')
+    ).then(catchStepError);
+  }
 
-  init: function n_init() {
-    this.mainWrapper = document.getElementById('main-wrapper');
-    this.transitionPromise = null;
+  var defer = Utils.Promise.defer();
 
-    currentPanel = null;
-  },
+  function onNavigationEnd() {
+    window.removeEventListener('navigation-transition-end', onNavigationEnd);
+    window.removeEventListener('load', onNavigationEnd);
+    defer.resolve(executeNavigationStep('afterEnter'));
+  }
 
-  setReady: function n_setReady() {
-    this.isReady = true;
-    nextQueuedPanel();
-  },
+  window.addEventListener('navigation-transition-end', onNavigationEnd);
+  window.addEventListener('load', onNavigationEnd); // simulate navigation end
 
-  /**
-   * Returns a boolean.
-   *
-   * Called with a first argument that is not an object, it returns whether the
-   * current panel is this first argument.
-   *
-   * Called with 2 arguments, the first one is not an object, the second one
-   * must be an object. It returns whether the current panel is the first
-   * argument, and for each property of the second argument, whether its value
-   * is equal to the value for the current panel's arguments' same property.
-   */
-  isCurrentPanel: function n_isCurrentPanel(panel, args) {
-    if (!currentPanel || !panel) {
-      return false;
-    }
+  return defer.promise.then(catchStepError);
+}
 
-    if (panel !== currentPanel.panel) {
-      return false;
-    }
+/**
+ * Used in `waitForSlideAnimation` to help solving race conditions.
+ * @type {Defer}
+ */
+var previousAnimationDefer;
+
+/**
+ * Wait for the end of an animation on `panelElement`.
+ *
+ * @param {Element} panelElement The element to watch.
+ * @returns {Promise} Resolves when the animation ends.
+ */
+function waitForSlideAnimation(panelElement) {
+  if (previousAnimationDefer) {
+    previousAnimationDefer.reject(new Error('A new animation started'));
+  }
 
-    if (!args) {
-      return true;
-    }
+  var defer = Utils.Promise.defer();
+  previousAnimationDefer = defer;
 
-    var currentArgs = currentPanel.args || {};
+  panelElement.addEventListener('animationend', function onAnimationEnd() {
+    this.removeEventListener('animationend', onAnimationEnd);
+    defer.resolve();
+  });
 
-    for (var arg in args) {
-      if (args[arg] != currentArgs[arg]) {
-        return false;
-      }
-    }
+  return defer.promise.then(() => previousAnimationDefer = null);
+}
```

Comment:
I think we should use Etienne's DOM Scheduler instead, but maybe in a future patch ?